lnwallet/chanfunding/psbt_assembler.go

```diff
@@ -0,0 +1,372 @@
+	// format. If the user supplied a base PSBT, we'll add the output to
+	// that one, otherwise we'll create a new one.
+	packet := i.BasePsbt
+	if packet == nil {
+		packet, err = psbt.New(nil, nil, 2, 0, nil)
+		if err != nil {
+			return "", 0, nil, fmt.Errorf("unable to create PSBT: "+
+				"%v", err)
+		}
+	}
+	packet.UnsignedTx.TxOut = append(packet.UnsignedTx.TxOut, out)
+	packet.Outputs = append(packet.Outputs, psbt.POutput{})
+
+	// Serialize the new or updated PSBT into the raw binary format.
+	var buf bytes.Buffer
+	err = packet.Serialize(&buf)
+	if err != nil {
+		return "", 0, nil, fmt.Errorf("unable to serialize PSBT: %v",
+			err)
+	}
+
+	return addr.EncodeAddress(), out.Value, buf.Bytes(), nil
+}
+
+// Verify makes sure the PSBT that is given to the intent has an output that
+// sends to the channel funding multisig address with the correct amount. A
+// simple check that there is a large enough input present is also performed.
+func (i *PsbtIntent) Verify(rawPsbt []byte) error {
+	// We want the PSBT in the raw binary format, base64 is only used when
+	// a direct user interaction is involved.
+	packet, err := psbt.NewFromRawBytes(bytes.NewReader(rawPsbt), false)
+	if err != nil {
+		return fmt.Errorf("error parseing PSBT: %v", err)
+	}
+
+	// Try to locate the channel funding multisig output.
+	_, expectedOutput, err := i.FundingOutput()
+	if err != nil {
+		return fmt.Errorf("funding output cannot be created: %v", err)
+	}
+	outputFound := false
+	for _, out := range packet.UnsignedTx.TxOut {
+		if reflect.DeepEqual(out, expectedOutput) {
+			outputFound = true
+		}
+	}
+	if !outputFound {
+		return fmt.Errorf("funding output not found in PSBT")
+	}
+
+	// At least one input needs to be specified. We don't want to dive
+	// deeper into input validation for the moment as that involves many
+	// intricacies. We see it as the input adder's job to make sure the
+	// amount is sufficient and that a change output is added.
+	if len(packet.UnsignedTx.TxIn) == 0 {
+		return fmt.Errorf("PSBT has no inputs")
+	}
+
+	i.PendingPsbt = packet
+	i.State = PsbtVerified
+	return nil
+}
+
+// Finalize makes sure the final PSBT that is given to the intent is fully valid
+// and signed but still contains the same UTXOs and outputs as the pending
+// transaction we previously verified. If everything checks out, the funding
+// manager is informed that the channel can now be opened and the funding
+// transaction be broadcast.
+func (i *PsbtIntent) Finalize(rawPsbt []byte) error {
+	// Parse the raw binary format and make sure the PSBT itself thinks it's
+	// finalized and ready to be broadcast.
+	packet, err := psbt.NewFromRawBytes(bytes.NewReader(rawPsbt), false)
+	if err != nil {
+		return fmt.Errorf("error parseing PSBT: %v", err)
+	}
+	err = psbt.MaybeFinalizeAll(packet)
+	if err != nil {
+		return fmt.Errorf("error finalizing PSBT: %v", err)
+	}
+	if !packet.IsComplete() {
```

Comment:
This can be omitted, as the first thing `Extract` does is ensure the packet is complete: https://github.com/btcsuite/btcutil/blob/master/psbt/extractor.go#L24